### PR TITLE
Fsm password persist

### PR DESCRIPTION
**Pull request by AleksandarFilipov**
Socketconnect state is persisted 
- naming in style `"ae-channel-service/apps/ae_socket_connector/log/2019-12-09 12/17/06.710671Z_channel_service_bob_test_close_on_chain_maliscous.log"`
- Allow test to specify log location (per peer), which allows subsequent test to reestablish using old state data
- In order to test the persisted data the persisted data is read upon reestablish.
- New test added where connect with persisted data will redirect to reestablish

### CODE REVIEW

File: apps/ae_socket_connector/lib/session_holder.ex [hunk 1]
```diff
@@ -8,18 +8,42 @@ defmodule SessionHolder do
...
         priv_key: priv_key,
+        connection_callbacks: connection_callbacks,
         color: color,
         # pid name, of the session holder, which is maintined over re-connect/re-establish
         pid_name: name
       }) do
-    GenServer.start_link(__MODULE__, {socket_connector_state, ae_url, network_id, priv_key, color}, name: name)
+    log_path = Map.get(log_config, :log_path, "log")
+    create_log_folder(log_path)
+    file_name_and_path = log_path <> "/" <> (Map.get(log_config, :log_file, generate_filename(name)))
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/session_holder.ex]: Not portable. Should use `:filename` functions.

File: apps/ae_socket_connector/lib/session_holder.ex [hunk 2]
```diff
@@ -8,18 +8,42 @@ defmodule SessionHolder do
...
+    case !File.exists?(file_name_and_path) do
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/session_holder.ex]: Using `if/else` would be easier to comprehend.

File: apps/ae_socket_connector/lib/session_holder.ex [hunk 3]
```diff
@@ -60,38 +84,104 @@ defmodule SessionHolder do
   end
 
   # Server
-  def init({%SocketConnector{} = socket_connector_state, ae_url, network_id, priv_key, color}) do
-    {:ok, pid} = SocketConnector.start_link(socket_connector_state, ae_url, network_id, color, self())
+  def init({socket_connector_state, ae_url, network_id, priv_key, connection_callbacks, file_name, :open, color}) do
+    :dets.open_file(String.to_atom(file_name), [type: :duplicate_bag])
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/session_holder.ex]: return value not checked

File: apps/ae_socket_connector/lib/session_holder.ex [hunk 4]
```diff
@@ -60,38 +84,104 @@ defmodule SessionHolder do
...
     Logger.debug("killing connector #{inspect(pid)}", ansi_color: color)
     Process.exit(pid, :normal)
   end
 
-  def fetch_state(pid) do
-    SocketConnector.request_state(pid)
+  # this is persent as a helper if you loose your channel id or password, then check this file.
+  defp persist(state, file) do
+    dets_state = %{time: (DateTime.utc_now |> DateTime.to_string()), state: state}
+    case :dets.insert(String.to_atom(file), {:connect, dets_state}) do
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/session_holder.ex]: The `dets` API is stateful, you need to keep a reference around to write to a file.

File: apps/ae_socket_connector/lib/session_holder.ex [hunk 5]
```diff
@@ -8,18 +8,42 @@ defmodule SessionHolder do
...
         ae_url: ae_url,
         network_id: network_id,
         priv_key: priv_key,
+        connection_callbacks: connection_callbacks,
         color: color,
         # pid name, of the session holder, which is maintined over re-connect/re-establish
         pid_name: name
       }) do
-    GenServer.start_link(__MODULE__, {socket_connector_state, ae_url, network_id, priv_key, color}, name: name)
+    log_path = Map.get(log_config, :log_path, "log")
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/session_holder.ex]: This path should be called `log`, but `data`. It contains data files.

**Review by tolbrino — DISMISSED** [review]

File: apps/ae_socket_connector/lib/session_holder.ex (hunk 4 above)
Comment (AleksandarFilipov) [reply apps/ae_socket_connector/lib/session_holder.ex]: Well sometimes you go with luck, turn out that the ref is the same as the filename. Thanks

File: apps/ae_socket_connector/lib/session_holder.ex (hunk 5 above)
Comment (AleksandarFilipov) [reply apps/ae_socket_connector/lib/session_holder.ex]: Did some attempt to address this. Confusing. Are you fine with _logging_ data to a data folder

Comment (AleksandarFilipov) [reply apps/ae_socket_connector/lib/session_holder.ex]: fixed

Comment (AleksandarFilipov) [reply apps/ae_socket_connector/lib/session_holder.ex]: Fixed

File: apps/ae_socket_connector/lib/client_runner.ex [hunk 6]
```diff
@@ -70,19 +70,19 @@ defmodule ClientRunner do
...
           session: state_channel_configuration,
-          role: role,
-          connection_callbacks: connection_callback(self(), color)
+          role: role
         },
+        log_config: log_config,
         ae_url: ae_url,
         network_id: network_id,
         priv_key: priv_key,
+        connection_callbacks: connection_callback(self(), color),
```
Comment (uwiger) [inline apps/ae_socket_connector/lib/client_runner.ex:85]:
```suggestion
        connection_callbacks: connection_callbacks(self(), color),
```

**Review by velzevur — APPROVED** [review]

**Review by uwiger — APPROVED** [review]

File: apps/ae_socket_connector/lib/client_runner.ex (hunk 6 above)
Comment (uwiger) [reply apps/ae_socket_connector/lib/client_runner.ex:85]: Obviously, the called function would need renaming as well.

**Pull request merged**